lib/ask/runtime/respondent_group_action.ex

```diff
@@ -291,18 +291,21 @@ defmodule Ask.Runtime.RespondentGroupAction do
     end
 
     respondent_ids = Enum.filter(entries, fn entry -> String.starts_with?(entry, "r") end)
-    phone_numbers = Enum.filter(entries, fn entry -> not String.starts_with?(entry, "r") end)
     phone_numbers_from_respondent_ids = phone_numbers_from_respondent_ids(survey, respondent_ids)
 
-    loaded_entries =
-      Enum.map(phone_numbers, fn phone_number -> %{phone_number: phone_number} end)
-      |> Enum.concat(phone_numbers_from_respondent_ids)
-
-    # Restore the initial entries order
+    # keeps the initial entries array in order
+    # OPTIMIZE: is it required to keep the order? it would be magnitudes faster if we didn't have to
     Enum.map(entries, fn entry ->
-      Enum.find(loaded_entries, fn %{phone_number: phone_number} = loaded_entry ->
-        entry == phone_number or entry == Map.get(loaded_entry, :hashed_number)
-      end)
+      if String.starts_with?(entry, "r") do
+        Enum.find(phone_numbers_from_respondent_ids, fn %{
```

Comment:
It could be interesting to use `Enum.delete` so on each iteration we'd have one less entry in the array... but I don't know how the Elixir runtime/Erlang VM react to inner list deletions?

Reply:
Alternatively if we wouldn't need to retain the original order, the whole function would become:

```elixir
    respondent_ids = Enum.filter(entries, fn entry -> String.starts_with?(entry, "r") end)
    phone_numbers = Enum.filter(entries, fn entry -> not String.starts_with?(entry, "r") end)
    phone_numbers_from_respondent_ids = phone_numbers_from_respondent_ids(survey, respondent_ids)

    phone_numbers
    |> Enum.map(fn phone_number -> %{phone_number: phone_number} end)
    |> Enum.concat(phone_numbers_from_respondent_ids)
    |> Enum.filter(&(!is_nil(&1)))
    |> Enum.uniq_by(fn %{phone_number: phone_number} -> keep_digits.(phone_number) end)
```